File: lib/features/analysis/domain/entities/budget.dart
enum BudgetPeriod { weekly, monthly, quarterly, annual, custom }
enum BudgetStatus { active, paused, completed, expired }

class Budget {
  final int id;
  final int userId;
  final int categoryId;
  final String name;
  final BudgetPeriod period;
  final BudgetStatus status;
  final int limitCents;
  final int spentCents;
  final String currency;
  final DateTime startDate;
  final DateTime? endDate;
  final bool isRollover;
  final int notifyAtPercentage;
  final DateTime createdAt;

  const Budget({
    required this.id,
    required this.userId,
    required this.categoryId,
    required this.name,
    required this.period,
    this.status = BudgetStatus.active,
    required this.limitCents,
    this.spentCents = 0,
    this.currency = 'USD',
    required this.startDate,
    this.endDate,
    this.isRollover = false,
    this.notifyAtPercentage = 80,
    required this.createdAt,
  });

  double get limit => limitCents / 100;
  double get spent => spentCents / 100;
  double get remaining => (limitCents - spentCents) / 100;
  double get usagePercentage => limitCents > 0 ? (spentCents / limitCents) * 100 : 0;
}

File: backend/src/domain/models/budget.py
import enum
from datetime import date, datetime

from sqlalchemy import Boolean, Date, DateTime, Enum, ForeignKey, Integer, String, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from src.config.database import Base


class BudgetPeriod(str, enum.Enum):
    WEEKLY = "weekly"
    MONTHLY = "monthly"
    QUARTERLY = "quarterly"
    ANNUAL = "annual"
    CUSTOM = "custom"


class BudgetStatus(str, enum.Enum):
    ACTIVE = "active"
    PAUSED = "paused"
    COMPLETED = "completed"
    EXPIRED = "expired"


class Budget(Base):
    __tablename__ = "budgets"

    id: Mapped[int] = mapped_column(primary_key=True)
    user_id: Mapped[int] = mapped_column(ForeignKey("users.id"), nullable=False)
    category_id: Mapped[int] = mapped_column(ForeignKey("categories.id"), nullable=False)
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    period: Mapped[BudgetPeriod] = mapped_column(Enum(BudgetPeriod), nullable=False)
    status: Mapped[BudgetStatus] = mapped_column(
        Enum(BudgetStatus), default=BudgetStatus.ACTIVE, nullable=False
    )
    limit_cents: Mapped[int] = mapped_column(nullable=False)
    spent_cents: Mapped[int] = mapped_column(default=0, nullable=False)
    currency: Mapped[str] = mapped_column(String(3), default="USD", nullable=False)
    start_date: Mapped[date] = mapped_column(Date, nullable=False)
    end_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    is_rollover: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    notify_at_percentage: Mapped[int] = mapped_column(default=80, nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False
    )

    user: Mapped["User"] = relationship(back_populates="budgets")
    category: Mapped["Category"] = relationship(back_populates="budgets")
